File: src/ClaudeUsageWatcher.Core/ThresholdTracker.cs
namespace ClaudeUsageWatcher.Core;

public enum ThresholdEvent
{
    Warning,
    Critical,
    Reset
}

public sealed class ThresholdTracker
{
    private readonly double _warnAt;
    private readonly double _criticalAt;
    private double? _prev;

    public ThresholdTracker(double warnAt = 80, double criticalAt = 95)
    {
        _warnAt = warnAt;
        _criticalAt = criticalAt;
    }

    public ThresholdEvent? Observe(double utilization)
    {
        if (_prev is not double prev)
        {
            // First observation only arms the tracker.
            _prev = utilization;
            return null;
        }

        _prev = utilization;

        if (prev < _criticalAt && utilization >= _criticalAt)
        {
            return ThresholdEvent.Critical;
        }

        if (prev < _warnAt && utilization >= _warnAt)
        {
            return ThresholdEvent.Warning;
        }

        if (prev >= 50 && utilization <= 10)
        {
            return ThresholdEvent.Reset;
        }

        return null;
    }
}

File: macos/Sources/UsageCore/ThresholdTracker.swift
import Foundation

public enum ThresholdEvent: Sendable {
    case warning
    case critical
    case reset
}

public final class ThresholdTracker {
    private let warnAt: Double
    private let criticalAt: Double
    private var prev: Double?

    public init(warnAt: Double = 80, criticalAt: Double = 95) {
        self.warnAt = warnAt
        self.criticalAt = criticalAt
    }

    public func observe(_ utilization: Double) -> ThresholdEvent? {
        guard let prev else {
            // First observation only arms the tracker.
            self.prev = utilization
            return nil
        }

        self.prev = utilization

        if prev < criticalAt && utilization >= criticalAt {
            return .critical
        }

        if prev < warnAt && utilization >= warnAt {
            return .warning
        }

        if prev >= 50 && utilization <= 10 {
            return .reset
        }

        return nil
    }
}
